src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/ClusterAwareMetricsContainer.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins.failover;
+
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.jdbc.JdbcConnection;
+import com.mysql.cj.jdbc.ha.plugins.ICurrentConnectionProvider;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ClusterAwareMetricsContainer implements IClusterAwareMetricsContainer {
+    // ClusterID, Metrics
+    private final static Map<String, ClusterAwareMetrics> clusterMetrics = new ConcurrentHashMap<>();
+    // Instance URL, Metrics
+    private final static Map<String, ClusterAwareMetrics> instanceMetrics = new ConcurrentHashMap<>();
+
+    private ICurrentConnectionProvider currentConnectionProvider;
+    private PropertySet propertySet;
+    private String clusterId = "";
+
+    public ClusterAwareMetricsContainer(ICurrentConnectionProvider currentConnectionProvider, PropertySet propertySet) {
+        this.currentConnectionProvider = currentConnectionProvider;
+        this.propertySet = propertySet;
+    }
+
+    public static void linkInstances(List<HostInfo> hosts, String clusterId) {
+        final ClusterAwareMetrics metrics = clusterMetrics.computeIfAbsent(
+            clusterId,
+            k -> new ClusterAwareMetrics());
+        for (HostInfo host : hosts) {
+            clusterMetrics.put(host.getHost(), metrics);
+        }
+    }
+
+    public void setClusterId(String clusterId) {
+        this.clusterId = !StringUtils.isNullOrEmpty(clusterId) ? clusterId.substring(0, clusterId.indexOf(':'))
+            : getCurrentConnUrl();
+    }
+
+    public void registerFailureDetectionTime(long timeMs) {
```

Comment:
I feel like there would be a suitable design pattern for handling these kind of behavior so we don't have to add the if statements for every method, but I can't recall on top of my head @sergiyv-bitquill do you have any suggestions?

Just a nit, would probably be fine like this. Just ugly.

Reply:
I'd propose a function that hide 
topologyMetrics.computeIfAbsent(clusterId, k -> new ClusterAwareTimeMetricsHolder("Topology Query"))
and accept lambda to be executed on ClusterAwareTimeMetricsHolder instance. But I don't think it's an improvement. It makes code less readable.